### PR TITLE
Move API key caching to authdb.

### PR DESCRIPTION
**Pull request by vadimberezniker**
The caching needs to be adjusted for subdomain handling and it makes more sense for the caching to be an internal detail of AuthDB.

**Related issues**: N/A

### CODE REVIEW

File: enterprise/server/backends/authdb/authdb.go [hunk 1]
```diff
@@ -37,23 +42,86 @@ const (
...
+	data         interfaces.APIKeyGroup
+	expiresAfter time.Time
+}
+
+// apiKeyGroupCache is a cache for API Key -> Group lookups. A single Bazel invocation
+// can generate large bursts of RPCs, each of which needs to be authed.
+// There's no need to go to the database for every single request as this data
+// rarely changes.
+type apiKeyGroupCache struct {
+	// Note that even though we base this off an LRU cache, every entry has a hard expiration
```
Comment (tylerwilliams) [inline enterprise/server/backends/authdb/authdb.go:68]: nit: 80 chars?

File: enterprise/server/backends/authdb/authdb.go [hunk 2]
```diff
@@ -37,23 +42,86 @@ const (
...
 
+	apiKeyGroupCache *apiKeyGroupCache
+
 	// Nil if API key encryption is not enabled.
 	apiKeyEncryptionKey []byte
 }
 
 func NewAuthDB(env environment.Env, h interfaces.DBHandle) (*AuthDB, error) {
 	adb := &AuthDB{env: env, h: h}
+	if *apiKeyGroupCacheTTL != time.Duration(0) {
```
Comment (tylerwilliams) [inline enterprise/server/backends/authdb/authdb.go:118]: if apiKeyGroupCacheTTL is a time.Duration you should just be able to compare it to 0?

File: enterprise/server/backends/authdb/authdb.go [hunk 3]
```diff
@@ -296,10 +372,22 @@ func (d *AuthDB) GetAPIKeyGroupFromAPIKey(ctx context.Context, apiKey string) (i
 		}
 		return nil, err
 	}
+	if d.apiKeyGroupCache != nil {
+		metrics.APIKeyLookupCount.With(prometheus.Labels{metrics.APIKeyLookupStatus: "cache_miss"}).Inc()
+		d.apiKeyGroupCache.Add(apiKey, akg)
+	} else {
+		metrics.APIKeyLookupCount.With(prometheus.Labels{metrics.APIKeyLookupStatus: "invalid_key"}).Inc()
```
Comment (tylerwilliams) [inline enterprise/server/backends/authdb/authdb.go:379]: it this in the right place? It seems like it will fire any time apiKeyGroupCache is nil.

**Review by tylerwilliams — APPROVED** [review]

File: enterprise/server/backends/authdb/authdb.go (hunk 1 above)
Comment (vadimberezniker) [reply enterprise/server/backends/authdb/authdb.go]: done

File: enterprise/server/backends/authdb/authdb.go (hunk 2 above)
Comment (vadimberezniker) [reply enterprise/server/backends/authdb/authdb.go]: done

File: enterprise/server/backends/authdb/authdb.go (hunk 3 above)
Comment (vadimberezniker) [reply enterprise/server/backends/authdb/authdb.go]: woops, this was also broken in the old code

**Pull request merged**